server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -129,6 +150,147 @@ public CompletionStage<Result> review(
+                  String selectedAddress =
+                      form.get(AddressCorrectionBlockView.SELECTED_ADDRESS_NAME);
+
+                  Optional<AddressSuggestion> suggestionMaybe =
+                      suggestionBuilder.build().stream()
+                          .filter(x -> x.getSingleLineAddress().equals(selectedAddress))
+                          .findFirst();
+
+                  Map<String, String> map = new HashMap<>(Map.of());
+
+                  if (suggestionMaybe.isPresent()) {
+                    AddressSuggestion suggestion = suggestionMaybe.get();
+                    Address address = suggestion.getAddress();
+                    AddressLocation location = suggestion.getLocation();
+
+                    map.put(addressQuestion.getStreetPath().toString(), address.getStreet());
+                    map.put(addressQuestion.getLine2Path().toString(), address.getLine2());
+                    map.put(addressQuestion.getCityPath().toString(), address.getCity());
+                    map.put(addressQuestion.getStatePath().toString(), address.getState());
+                    map.put(addressQuestion.getZipPath().toString(), address.getZip());
+                    map.put(
+                        addressQuestion.getLatitudePath().toString(),
+                        location.getLatitude().toString());
+                    map.put(
+                        addressQuestion.getLongitudePath().toString(),
+                        location.getLongitude().toString());
+                    map.put(
+                        addressQuestion.getWellKnownIdPath().toString(),
+                        location.getWellKnownId().toString());
+                    map.put(
+                        addressQuestion.getCorrectedPath().toString(),
+                        CorrectedAddressState.CORRECTED.toString());
+                  } else if (selectedAddress.equals(
+                      AddressCorrectionBlockView.USER_KEEPING_ADDRESS_VALUE)) {
+                    map.put(
+                        addressQuestion.getCorrectedPath().toString(),
+                        CorrectedAddressState.AS_ENTERED_BY_USER.toString());
+                  } else {
+                    map.put(
+                        addressQuestion.getCorrectedPath().toString(),
+                        CorrectedAddressState.FAILED.toString());
+                  }
+
+                  ImmutableMap<String, String> formData = cleanForm(map);
+                  applicantService.stageAndUpdateIfValid(applicantId, programId, blockId, formData);
+                }
+
+                Optional<String> nextBlockIdMaybe =
+                    roApplicantProgramService.getInProgressBlockAfter(blockId).map(Block::getId);
+                // No next block so go to the program review page.
+                if (nextBlockIdMaybe.isEmpty()) {
+                  return redirect(
+                      routes.ApplicantProgramReviewController.review(applicantId, programId));
```

Comment:
@bion I'm having one issue here where if I redirect to the review page or were I to get redirected back to the edit page of the same block the data seems to be cached. If I refresh the page manually (F5) it reloads with current data. Is there something I can call to refresh that cached data?

Reply:
Could you call `renderErrorOrRedirectToNextBlock` here instead of implementing the redirect logic directly?

Reply:
The caching problem I had I fixed. I wasn't using the new ReadOnlyApplicantProgramService returned from stageAndUpdateIfValid so I was looking at stale data. 

Switched to renderErrorOrRedirectToNextBlock.